pkg/driver/controller_modify_volume.go

```diff
@@ -17,8 +20,147 @@ const (
 	ModificationKeyIOPS = "iops"
 
 	ModificationKeyThroughput = "throughput"
+
+	modifyVolumeRequestHandlerTimeout = 2 * time.Second
 )
 
+type modifyVolumeRequest struct {
+	newSize           int64
+	modifyDiskOptions cloud.ModifyDiskOptions
+}
+
+type modifyVolumeRequestHandler struct {
+	volumeID string
+	// Merged request from the requests that have been accepted for the volume
+	mergedRequest *modifyVolumeRequest
+	// Channel for sending requests to the Timer thread for the volume
+	requestChan chan *modifyVolumeRequest
+	// Channel for sending errors of ec2 ModifyVolume API call to the CSI Driver main thread
+	errChan chan error
+	// Channel for sending the actual volume size after the ec2 ModifyVolume API call to the CSI Driver main thread
+	resultChan chan int64
+	mux sync.RWMutex
+}
+
+type modifyVolumeManager struct {
+	// Map of volume ID to modifyVolumeRequestHandler
+	requestHandlerMap map[string]*modifyVolumeRequestHandler
+	mux               *sync.RWMutex
+}
+
+func newModifyVolumeManager() *modifyVolumeManager {
+	return &modifyVolumeManager{
+		requestHandlerMap: make(map[string]*modifyVolumeRequestHandler),
+		mux:               &sync.RWMutex{},
+	}
+}
+
+func (m *modifyVolumeManager) deleteRequestHandler(volumeID string) {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+	delete(m.requestHandlerMap, volumeID)
+}
+
+// This function validates the new request against the merged request for the volume.
+// If the new request has a volume property that's already included in the merged request and its value is different from that in the merged request,
+// this function will return an error and the new request will be rejected.
+func (h *modifyVolumeRequestHandler) validateModifyVolumeRequest(r *modifyVolumeRequest) error {
+	h.mux.RLock()
+	defer h.mux.RUnlock()
+	if r.newSize != 0 && h.mergedRequest.newSize != 0 && r.newSize != h.mergedRequest.newSize {
+		return fmt.Errorf("Different size was requested by a previous request")
+	}
+	if r.modifyDiskOptions.IOPS != 0 && h.mergedRequest.modifyDiskOptions.IOPS != 0 && r.modifyDiskOptions.IOPS != h.mergedRequest.modifyDiskOptions.IOPS {
+		return fmt.Errorf("Different IOPS was requested by a previous request")
+	}
+	if r.modifyDiskOptions.Throughput != 0 && h.mergedRequest.modifyDiskOptions.Throughput != 0 && r.modifyDiskOptions.Throughput != h.mergedRequest.modifyDiskOptions.Throughput {
+		return fmt.Errorf("Different throughput was requested by a previous request")
+	}
+	if r.modifyDiskOptions.VolumeType != "" && h.mergedRequest.modifyDiskOptions.VolumeType != "" && r.modifyDiskOptions.VolumeType != h.mergedRequest.modifyDiskOptions.VolumeType {
+		return fmt.Errorf("Different volume type was requested by a previous request")
+	}
+	return nil
+}
+
+func (h *modifyVolumeRequestHandler) mergeModifyVolumeRequest(r *modifyVolumeRequest) {
+	h.mux.Lock()
```

Comment:
validateModifyVolumeRequest and mergeModifyVolumeRequest need to happen in the same lock-context, otherwise you create a window where an incompatible request can race in after the validate, but before the merge, and we won't notice.

Reply:
Done. See the latest ``addModifyVolumeRequest`` and ``startVolumeTimer`` methods.